### DISCUSSION TITLE
How to specify standalone grouped up backends for either completion after idle or 'company-complete'

### DISCUSSION THREAD

**Discussion by IceAsteroid** [company-mode/company-mode#1480] [Q&A] [answered]
I know that both 'auto completion after idle' and `company-complete` both use the variable `company-backends`.

Now, I have a use case where I need only to show `company-dabbrev-code` for 'auto completion after idle' and have a keybinding bound to `company-complete` to show candidates provided by a list of grouped-up backends specified in `company-backend` variable.

The solution that I imagined is either
* be able to specific standalone grouped backends for `company-complete` & Company-backends for "completion after idle".
* be able to specific standalone grouped backends for "completion after idle" & `company-backends` for keybinding bound to `company-complete`.

How could it be done?

------------
I have a workaround but it's not versatile because I found no way to group backends for either of the ways I described above.

```
    (add-hook 'haskell-mode-hook (lambda ()
				   ;; (setq-local company-backends
				   ;; 	       '((company-dabbrev-code
				   ;; 		  company-capf)))
				   ;; 		  :separate)))
				   (setq-local company-backends
					       '((company-dabbrev-code)))
				   (keymap-local-set "M-<tab>" (lambda ()
								 (interactive)
								 (company-abort)
								 (company-begin-backend 'company-capf)))
				   (interactive-haskell-mode 1)
				   (eglot-ensure)))
```

I've searched & browsed throught the issue & discussion pages & user manual for this, but found no answer.

**Comment by dgutov** [comment]
Hi!

This is non-trivial to do "properly", but some hacks might be within easy each. Let's try to establish the requirements.

Normally, when the user invokes `company-complete` or `company-complete-common` or `company-next-completion`, the completion turns from "idle" into "manual" one. But would you want the set of the backends used to switch as well? After each of these commands, or specific ones?

**Reply by IceAsteroid** [reply]
Good day, @dgutov. Sorry for the delay, I finally got some time.
 
Did you mean that the switch is done by the command `company-other-backend`, and it uses the variable `company-backends` to switch through?

If so, that's indeed a use case.

However, even when I set the `company-backends` lexically within a keybinding bound to `company-backend`, but invoking 'company-other-backend' will still not use the lexical variable as the list, as follows:

```
(keymap-local-set "M-<tab>" (lambda ()
						        (interactive)
							(company-abort)
						        (let ((company-backends '(company-dabbrev-code
											          company-capf)))
								   (company-complete))))
```

The message outputs "No other backend". In case to figure out, I've also set several backends outside the lexical, switching is fine with "completion after idle", I don't know what's going on.

-------------------------

And I've managed to group backends into one with `:separate` keyword for a keybinding, with lexical binding, like the following:

```
(keymap-local-set "M-<tab>" (lambda ()
							(interactive)
							(company-abort)
							(let ((company-backends '((company-dabbrev-code
											            company-capf :separate))))
								   (company-complete))))
```

It's nice if there's a way for `company-other-backend` to be used to switch a list of backends for a keybinding.

In my use cases, I'm rarely using it, I'm mostly using the "grouped backends with :separate" feature.

**Reply by dgutov** [reply]
> Did you mean that the switch is done by the command company-other-backend, and it uses the variable company-backends to switch through?

I didn't mean that command in particular (yes, it's a separate command), but it can be added to the mentioned list, yes.

> However, even when I set the company-backends lexically within a keybinding bound to company-backend, but invoking 'company-other-backend' will still not use the lexical variable as the list, as follows:

You are not using `company-other-backend` in both snippets, though. They both include calls to `company-complete` - it's a different command.

> It's nice if there's a way for company-other-backend to be used to switch a list of backends for a keybinding.

That command switches to the next configured backend in `company-backends` (the one after the currently used one). Is that the effect you were looking for?

**Reply by IceAsteroid** [reply]
I meant, to have the standalone backend list to switch, specifically for the completion via a keybinding, itself separated from the "company-backends" that's widely used by things like the "completion after idle".
 
> Normally, when the user invokes company-complete or company-complete-common or company-next-completion, the completion turns from "idle" into "manual" one. **But would you want the set of the backends used to switch as well? After each of these commands, or specific ones?**

But in my attempt of the configuration, I wasn't able to successfully achieve it. I'm not very good at elisp, I'm sorry.

I was invoking the `company-other-backend` during the popup of completion tooltip launched by the keybinding that's defined to have a let expression to have the lexical scope "company-backends" bound in it, and `company-complete` in it.

Like what I posted:

> ```
> (keymap-local-set "M-<tab>" (lambda ()
> 						        (interactive)
> 							(company-abort)
> 						        (let ((company-backends '(company-dabbrev-code
> 											          company-capf)))
> 								   (company-complete))))
> ```

But `company-other-backend` outputs: `No other backend`.
 
I'm sorry for the unclear explanation.

**Reply by dgutov** [reply]
> I was invoking the company-other-backend during the popup of completion tooltip launched by the keybinding that's defined to have a let expression to have the lexical scope "company-backends" bound in it, and company-complete in it.

Indeed, the let-binding like in your example set a value that is restored at the end of a command. So any later commands that look up `company-backends` again won't see that value.

If you want to use it with `company-other-backend`, you could create a similar wrapper command for it, which would also temporarily set `company-backends` to a value like the above.

**Reply by IceAsteroid** [reply]
@dgutov, fine, thank you for the patience :)

**Comment by dgutov** [comment, marked as answer]
Speaking of your workaround solution, the `backend` argument can be a list too.

So you can try calling `(company-begin-backend '(company-capf company-dabbrev-code))`, for example.

**Reply by IceAsteroid** [reply]
Hi, dgutov, grettings!

Thanks for the solution. It looks nice. I'm gonna respond your first comment and try out the solution in the second command later.
